### PR TITLE
[VR-10714] Async implementation for the project attributes endpoints

### PR DESCRIPTION
**Pull request by anandjakhaniya**
**Jira Ticket:** https://vertaai.atlassian.net/browse/VR-10714

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/KeyValueHandler.java [hunk 1]
```diff
@@ -116,7 +116,8 @@ public KeyValueHandler(Executor executor, FutureJdbi jdbi, String fieldType, Str
                             .findOne()
                             .ifPresent(
                                 present -> {
-                                  throw new AlreadyExistsException("Key already exists");
+                                  throw new AlreadyExistsException(
+                                      "Key " + kv.getKey() + " already exists");
```
Comment (ghost) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/KeyValueHandler.java:120]: nice 👍🏼

File: backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java [hunk 2]
```diff
@@ -63,27 +66,80 @@ public void updateProjectDescription(
   public void addProjectAttributes(
       AddProjectAttributes request,
       StreamObserver<AddProjectAttributes.Response> responseObserver) {
-    super.addProjectAttributes(request, responseObserver);
+    try {
+      final var response =
```
Comment (ghost) [inline backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java]: I like to name future variables with `future` in front, like `futureResponse`, for my own sanity; I think it helps.

**Review by ghost — APPROVED** [review]

File: backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java [hunk 3]
```diff
@@ -48,7 +50,7 @@ public FutureProjectDAO(Executor executor, FutureJdbi jdbi, UAC uac) {
     tagsHandler = new TagsHandler(executor, jdbi, entityName);
   }
 
-  public InternalFuture<Void> deleteAttributes(DeleteExperimentRunAttributes request) {
+  public InternalFuture<Void> deleteAttributes(DeleteProjectAttributes request) {
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java]: Unrelated change

File: backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java [hunk 4]
```diff
@@ -146,4 +158,9 @@ public FutureProjectDAO(Executor executor, FutureJdbi jdbi, UAC uac) {
             },
             executor);
   }
+
+  public InternalFuture<Project> getPreAccessProjectById(String projectId) {
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java]: What is this? Why is this public? Assuming `preAccess` means without authorization check, it doesn't make sense to be public, as per our guidelines https://github.com/VertaAI/docs/tree/main/engineering-practices/api-implementation

**Review by conradoverta — CHANGES_REQUESTED** [review]

File: backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java [hunk 5]
```diff
@@ -63,27 +66,80 @@ public void updateProjectDescription(
...
-    super.deleteProjectAttributes(request, responseObserver);
+    try {
+      final var response =
+          futureProjectDAO
+              .deleteAttributes(request)
+              .thenCompose(
+                  unused -> futureProjectDAO.getPreAccessProjectById(request.getId()), executor)
+              .thenApply(
+                  project ->
+                      DeleteProjectAttributes.Response.newBuilder().setProject(project).build(),
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java]: @hmacdonald-verta could you check with @convoliution that all of these APIs returning `Project` actually need them to return `Project`? I think this data is just discarded by the caller, so we could save some round trips here.

File: backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java (hunk 3 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java]: This Pr is for the Project entity and this method is foe delete project attributes not the experiment run so this is the valid change as per PR heading.

Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java]: No, it is not. It is a bad copy and paste from before. I'm _really_ trying to remove the bad practice of adding a bunch of unrelated changes to the same PR as it makes software much less robust.

File: backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java (hunk 5 above)
Comment (convoliution) [reply backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java]: The only time the client uses the `Project` returned in the response is in

- `/getProjectById`
- `/getProjectByName`
- `/createProject`

Erm...in fact, the python client doesn't even expose mutating attributes on projects, anyway.

Comment (convoliution) [reply backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java]: It doesn't look like the Scala client uses the returned `Project` for these endpoints, either.

Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java]: @conradoverta then should we ignore to feel project in the response??

File: backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java (hunk 3 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java]: Yes. Fixed on the base PR

File: backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java]: Fixed

File: backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/KeyValueHandler.java (hunk 1 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/KeyValueHandler.java:120]: Thanks

File: backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java (hunk 5 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java]: Yes. Let's also put up a PR to remove that from the return completely for clarity too.

Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/project/FutureProjectServiceImpl.java]: removed code for now in PR: https://github.com/VertaAI/modeldb/pull/2129

File: backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java (hunk 4 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/project/FutureProjectDAO.java]: removed code for now in PR: #2129

File: backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/KeyValueHandler.java [hunk 6]
```diff
@@ -171,4 +172,40 @@ public KeyValueHandler(Executor executor, FutureJdbi jdbi, String fieldType, Str
...
+            unused ->
+                // Update into KV table
+                jdbi.useHandle(
+                    handle -> {
+                      handle
+                          .createUpdate(
+                              "Update "
+                                  + getTableName()
+                                  + " SET kv_key=:key, kv_value=:value, value_type=:type "
+                                  + " where entity_name=:entity_name and field_type=:field_type and kv_key=:key and "
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/KeyValueHandler.java:199]: What is supposed to happen if the given key doesn't exist?

Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/KeyValueHandler.java:199]: Not sure. but what is expected here?? should we throw error if not exists Or we will just add it in database??

Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/KeyValueHandler.java:199]: On old flow if key was not exists then we are logged it as a new entry.

Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/KeyValueHandler.java:199]: No strong preference. We might end up removing this update since ERs don't have them, so they are probably not strictly necessary. That will come in later. Add a TODO for us to revisit

Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/KeyValueHandler.java:199]: Added TODO on the method

**Review by conradoverta — APPROVED** [review]
